src/engine/PDKind.cc

```diff
@@ -0,0 +1,341 @@
+﻿/*
+* Copyright (c) 2023-2024, Stepan Henrych <stepan.henrych@gmail.com>
+*
+* SPDX-License-Identifier: MIT
+*/
+
+#include "PDKind.h"
+
+#include "Common.h"
+#include "MainSolver.h"
+#include "ModelBasedProjection.h"
+#include "PTRef.h"
+#include "QuantifierElimination.h"
+#include "TermUtils.h"
+#include "TransformationUtils.h"
+#include "transformers/BasicTransformationPipelines.h"
+#include "transformers/SingleLoopTransformation.h"
+#include <memory>
+#include <queue>
+#include <set>
+#include <tuple>
+
+VerificationResult PDKind::solve(ChcDirectedHyperGraph const & graph) {
+    auto pipeline = Transformations::towardsTransitionSystems();
+    auto transformationResult = pipeline.transform(std::make_unique<ChcDirectedHyperGraph>(graph));
+    auto transformedGraph = std::move(transformationResult.first);
+    auto translator = std::move(transformationResult.second);
+    if (transformedGraph->isNormalGraph()) {
+        auto normalGraph = transformedGraph->toNormalGraph();
+        auto res = solve(*normalGraph);
+        return computeWitness ? translator->translate(std::move(res)) : std::move(res);
+    }
+    return VerificationResult(VerificationAnswer::UNKNOWN);
+}
+
+VerificationResult PDKind::solve(ChcDirectedGraph const & system) {
+    if (isTrivial(system)) {
+        return solveTrivial(system);
+    }
+    if (isTransitionSystem(system)) {
+        auto ts = toTransitionSystem(system);
+        auto res = solveTransitionSystem(*ts);
+        return translateTransitionSystemResult(res, system, *ts);
+    }
+    SingleLoopTransformation transformation;
+    auto[ts, backtranslator] = transformation.transform(system);
+    assert(ts);
+    auto res = solveTransitionSystem(*ts);
+    return translateTransitionSystemResult(res, system, *ts);
```

Comment:
```suggestion
    return computeWitness ? backtranslator->translate(res) : VerificationResult(res.answer);
```

This is actually a special case and the backtranslator from `SingleLoopTransformation` takes proper care mapping the result back on the graph. (See, e.g., IMC engine.)